src/transformStateWithClones.js

```diff
@@ -7,7 +7,51 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateCopy = { ...state };
+  const results = [];
+
+  for (const action of actions) {
+    const newRecord = { ...stateCopy };
+
+    switch (action.type) {
+      case 'addProperties':
+        addData(newRecord, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeProperties(newRecord, action.keysToRemove);
+        break;
+
+      case 'clear':
+        clear(newRecord);
+        break;
+
+      default:
+        break;
+    }
+
+    stateCopy = newRecord;
```

Comment:
According to the additional prompt rules, it is better to push a copy of the object to the array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed.